lib/internal/bootstrap/node.js

```diff
@@ -384,17 +403,34 @@ function setupGlobalProxy() {
 
 function setupBuffer() {
   const { Buffer } = NativeModule.require('buffer');
+  const { cjsContext } = NativeModule;
+  const { getStackSourceName } = internalBinding('contextify');
   const bufferBinding = internalBinding('buffer');
+  const experimentalModules = getOptionValue('--experimental-modules');
 
   // Only after this point can C++ use Buffer::New()
   bufferBinding.setBufferPrototype(Buffer.prototype);
   delete bufferBinding.setBufferPrototype;
   delete bufferBinding.zeroFill;
 
+  cjsContext.Buffer = Buffer;
+  const getter = deprecate(
+    () => cjsContext.Buffer,
+    'global.Buffer is deprecated in ECMAScript Modules. ' +
+        'Please use `import { Buffer } from \'buffer\'` instead.',
+    'DEP0126');
+
   Object.defineProperty(global, 'Buffer', {
```

Comment:
this could cause some pretty serious performance regressions if people are generally used to accessing Buffer without overhead. are there some tests we can run to check if this is okay on that front?

Reply:
Note that only `global.Buffer` is affected here, not `Buffer`. `global.Buffer` access in a hotpath seems unlikely (most people do just `Buffer`), but it could be worth verifying that certainly.

Reply:
Unless the global Buffer is shadowed per module, a bare global variable reference is supposed to exercise a getter. Is it shadowed?

Reply:
Yes we shadow the global reference in CommonJS with the direct Buffer and process references, with the shadowed reference being synchronized with the global.Buffer and global.process versions. This ensures the fast path in the majority of use cases.